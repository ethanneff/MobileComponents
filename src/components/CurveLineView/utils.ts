import Victor from "victor";
import Errors from "../../commons/Errors";
import { Point2 } from "../../commons/Utils/Maths/geometry/Point";

export function getMidPointPosition(
  startPosition: Point2,
  endPosition: Point2
) {
  return new Victor(
    (startPosition.x + endPosition.x) / 2,
    (startPosition.y + endPosition.y) / 2
  );
}

// Vector stuff

type Scalar = number;

function vAdd(v1: Point2, v2: Point2): Victor {
  return new Victor(v1.x + v2.x, v1.y + v2.y);
}
function vTimes(scalar: Scalar, vector: Victor): Victor {
  return new Victor(vector.x * scalar, vector.y * scalar);
}
const vLength = ({ x, y }: Victor) => Math.sqrt(x * x + y * y);

/**
 *
 * @param v
 * @returns a vector orthogonal to `v`
 */
function vPerpendicular({ x, y }: Victor): Victor {
  if (y === 0 && x === 0) {
    throw new Error(Errors.INVALID_ORTHOGANAL_VECTOR);
  } else if (y === 0) {
    return new Victor(-y / x, 1);
  }

  return new Victor(1, -x / y);
}

function vNormalize(v: Victor): Victor {
  return vTimes(1 / vLength(v), v);
}

function toRadians(angle: number) {
  return angle * (Math.PI / 180);
}

export const calculateControlPoints = (
  startPosition: Point2,
  midPointPosition: Point2,
  targetZoneAngle: number,
  sensorAngle: number,
  tolerance: number
) => {
  const targetZoneAngleInner = targetZoneAngle - tolerance / 2;
  const targetZoneAngleOuter = targetZoneAngle + tolerance / 2;

  return {
    controlPointOuter: getControlPoint(
      startPosition,
      midPointPosition,
      targetZoneAngleOuter
    ),
    controlPointInner: getControlPoint(
      startPosition,
      midPointPosition,
      targetZoneAngleInner
    ),
    controlPointLine: getControlPoint(
      startPosition,
      midPointPosition,
      sensorAngle
    )
  };
};

const getControlPoint = (start: Point2, midpoint: Point2, angle: number) => {
  const vectorA = new Victor(start.x - midpoint.x, start.y - midpoint.y);
  const lengthA = vLength(vectorA);

  const lengthB =
    (lengthA * Math.sin(toRadians(angle))) / Math.cos(toRadians(angle));

  const vectorPerpendicular = vNormalize(vPerpendicular(vectorA));

  const vectorPoint = vTimes(lengthB, vectorPerpendicular);
  const controlPoint = vAdd(midpoint, vectorPoint);

  return controlPoint;
};
