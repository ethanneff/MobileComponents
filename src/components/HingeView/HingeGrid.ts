import { createPoint2 } from "../../commons/Utils/Maths/geometry/Point";

export const HingeGridPositionPoints = {
  centerLeft: createPoint2(550, 880),
  centerRight: createPoint2(1050, 880),
  centerMiddle: createPoint2(800, 880),
  topLeft: createPoint2(550, 690),
  topMiddle: createPoint2(800, 690),
  topRight: createPoint2(1050, 690),
  bottomMiddle: createPoint2(800, 1110),
  bottomRight: createPoint2(1050, 1110),
  bottomLeft: createPoint2(550, 1110)
};

export type HingeGridPosition = keyof typeof HingeGridPositionPoints;
