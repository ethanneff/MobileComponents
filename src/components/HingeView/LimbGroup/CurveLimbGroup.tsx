import React from "react";
import { G } from "react-native-svg";
import Theme from "../../../commons/Theme";
import { Point2 } from "../../../commons/Utils/Maths/geometry/Point";
import { describeBezier } from "../../../commons/Utils/SvgUtils";
import Curve from "../Curve";
import Hinge from "../Hinge";

interface CurveLimbGroupProps {
  startHingePosition: Point2;
  endHingePosition: Point2;
  controlPoint: Point2;
  startHingeId?: string;
  endHingeId?: string;
  scaledDown?: boolean;
}

const styles = (scaledDown?: boolean) => ({
  fixedHinge: {
    fillColor: Theme.color.neutral300,
    strokeColor: Theme.color.neutral300,
    radius: scaledDown ? 16 : 22,
    strokeWidth: 12
  },
  fixedLimb: {
    strokeColor: Theme.color.neutral300,
    strokeWidth: scaledDown ? 18 : 27
  }
});

export default function CurveLimbGroup({
  startHingePosition,
  endHingePosition,
  controlPoint,
  startHingeId,
  endHingeId,
  scaledDown
}: CurveLimbGroupProps) {
  const curveData = describeBezier(
    startHingePosition,
    controlPoint,
    endHingePosition
  );
  return (
    <G>
      <Curve curveData={curveData} />
      <Hinge
        id={startHingeId}
        position={startHingePosition}
        radius={styles(scaledDown).fixedHinge.radius}
        strokeColor={styles(scaledDown).fixedHinge.strokeColor}
        strokeWidth={styles(scaledDown).fixedHinge.strokeWidth}
        fillColor={styles(scaledDown).fixedHinge.fillColor}
      />
      <Hinge
        id={endHingeId}
        position={endHingePosition}
        radius={styles(scaledDown).fixedHinge.radius}
        strokeColor={styles(scaledDown).fixedHinge.strokeColor}
        strokeWidth={styles(scaledDown).fixedHinge.strokeWidth}
        fillColor={styles(scaledDown).fixedHinge.fillColor}
      />
    </G>
  );
}
