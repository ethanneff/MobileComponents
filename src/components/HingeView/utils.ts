import { Point2 } from "../../commons/Utils/Maths/geometry/Point";
import { polarToCartesian } from "../../commons/Utils/Trigo";

export const LIMB_LENGTH = 400;
export const SCALED_DECREMENT = 96;
export const TARGET_PATH_RADIUS = 450;
export const CENTERED_AND_SCALED_LIMB_LENGTH = LIMB_LENGTH - SCALED_DECREMENT;
export const CENTERED_AND_SCALED_TARGET_PATH_LENGTH =
  TARGET_PATH_RADIUS - SCALED_DECREMENT;

export const angleVectors = (
  angle: number,
  startPosition: Point2,
  scaledDown?: boolean
) =>
  polarToCartesian(
    startPosition.x,
    startPosition.y,
    scaledDown ? CENTERED_AND_SCALED_LIMB_LENGTH : LIMB_LENGTH,
    angle
  );
