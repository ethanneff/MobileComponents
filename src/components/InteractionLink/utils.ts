import DeviceInfo from "react-native-device-info";

export type InteractionLinkCommunicationType = "email" | "phone" | "text";

export type InteractionLinkContent =
  | {
      message: string;
      type?: InteractionLinkCommunicationType;
    }
  | {
      message: string;
      type?: "navigate";
      route: string;
    };

export type InteractionLinkContentType = InteractionLinkContent["type"];

export const hasInteraction = (type?: InteractionLinkContentType) =>
  type === "navigate" ||
  ((type === "email" || type === "phone") &&
    !DeviceInfo.isEmulatorSync() &&
    DeviceInfo.getBrand() !== "Amazon");

export const interactionLink = (
  message: string,
  type?: InteractionLinkContentType
) =>
  type === "phone"
    ? `tel:${message}`
    : type === "email"
    ? `mailto:${message}`
    : type === "text"
    ? `sms:${message}`
    : message;
