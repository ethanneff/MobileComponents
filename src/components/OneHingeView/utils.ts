import { AngleTransformation } from "./interfaces";

/**
 * Returns appropriate angle transforming function
 * depending on the exercise being rendered.
 */
export function getAngleTransformation(transformation: AngleTransformation) {
  switch (transformation) {
    case AngleTransformation.SeatedHamstringStretch:
    case AngleTransformation.Woodpecker:
      return (angle: number) => 360 - angle;
    case AngleTransformation.SideBend:
      return (angle: number) => 270 - angle;
    case AngleTransformation.Clamshell:
      return (angle: number) => 270 + angle;
    case AngleTransformation.SidelyingRotation:
      return (angle: number) => 180 - angle;
    case AngleTransformation.FireHydrant:
    case AngleTransformation.InternalHipRotation:
    case AngleTransformation.ExternalHipRotation:
    case AngleTransformation.SidePlankLeft:
    case AngleTransformation.SidePlankRight:
    case AngleTransformation.SidePlank:
      return (angle: number) => 90 - angle;
  }
}
