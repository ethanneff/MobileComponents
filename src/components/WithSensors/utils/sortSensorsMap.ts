import { SensorMap } from "../../../commons/Store/Sensor";

export function sortSensorsMap(map: SensorMap) {
  return Object.values(map || {}).sort((left, right) => {
    if (!left || !right || !left.id || !right.id) {
      return 0;
    }
    if (left.id < right.id) {
      return -1;
    }
    if (left.id > right.id) {
      return 1;
    }
    return 0;
  });
}
