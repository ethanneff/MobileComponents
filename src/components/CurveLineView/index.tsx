import React, { memo, useEffect } from "react";
import Svg from "react-native-svg";
import { mod } from "../../commons/Utils/Maths";
import { Point2 } from "../../commons/Utils/Maths/geometry/Point";
import CurveTargetZone from "../HingeView/CurveTargetZone";
import CurveLimbGroup from "../HingeView/LimbGroup/CurveLimbGroup";
import { calculateControlPoints, getMidPointPosition } from "./utils";

interface TargetZoneProps {
  targetZoneAngle: number;
  tolerance: number;
}

interface ComponentProps {
  canvasSize: number;
  sensorAngle: number;
  isInTargetZone: boolean;
  startPosition: Point2;
  endPosition: Point2;
  onIsInTargetZoneChange(val: boolean): void;
}

type Props = TargetZoneProps & ComponentProps;

const normalizedAngleCalculation = (angle: number) => mod(angle - 89, 178) - 89;

/**
 * @description for rendering the curve line and target zone
 * @param {number} targetZoneAngle needs to be a value that is above -90 and below 90
 */

export default memo(function CurveLineView({
  canvasSize,
  sensorAngle,
  targetZoneAngle,
  tolerance,
  isInTargetZone,
  onIsInTargetZoneChange,
  startPosition,
  endPosition
}: Props) {
  const normalizedTargetZoneAngle = normalizedAngleCalculation(targetZoneAngle);

  const normalizedAngle = normalizedAngleCalculation(sensorAngle);

  const controlPointPosition = getMidPointPosition(startPosition, endPosition);

  const vectorPoints = calculateControlPoints(
    startPosition,
    controlPointPosition,
    normalizedTargetZoneAngle,
    normalizedAngle,
    tolerance
  );

  const newIsInTargetZone = calculateIsInTargetZone(
    normalizedTargetZoneAngle,
    tolerance,
    normalizedAngle
  );

  useEffect(() => {
    if (newIsInTargetZone !== isInTargetZone) {
      onIsInTargetZoneChange(newIsInTargetZone);
    }
  }, [newIsInTargetZone, isInTargetZone, onIsInTargetZoneChange]);
  return (
    <Svg width={canvasSize} height={canvasSize} viewBox={"0 0 1600 1600"}>
      <CurveTargetZone
        isInTargetZone={isInTargetZone}
        startHingePosition={startPosition}
        endHingePosition={endPosition}
        controlPointOuter={vectorPoints.controlPointOuter}
        controlPointInner={vectorPoints.controlPointInner}
      />
      <CurveLimbGroup
        startHingePosition={startPosition}
        endHingePosition={endPosition}
        controlPoint={vectorPoints.controlPointLine}
      />
    </Svg>
  );
});

const calculateIsInTargetZone = (
  targetZoneAngle: number,
  targetZoneTolerance: number,
  sensorAngle: number
) => {
  const delta = targetZoneTolerance / 2;
  return (
    targetZoneAngle - delta <= sensorAngle &&
    sensorAngle <= targetZoneAngle + delta
  );
};
