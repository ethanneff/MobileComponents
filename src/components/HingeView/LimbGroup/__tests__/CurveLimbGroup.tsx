import { render } from "@testing-library/react-native";
import React from "react";
import { getMidPointPosition } from "../../../CurveLineView/utils";
import { HingeGridPositionPoints } from "../../HingeGrid";
import CurveLimbGroup from "../CurveLimbGroup";

jest.mock("../../../../commons/Store/selectors", () => {
  return {
    useRootSelector: () => jest.fn()
  };
});

describe("CurveLimbGroup", () => {
  it("renders correctly", () => {
    const startPosition = HingeGridPositionPoints.centerLeft;
    const endPosition = HingeGridPositionPoints.centerRight;
    const midPoint = getMidPointPosition(startPosition, endPosition);
    const { container } = render(
      <CurveLimbGroup
        startHingePosition={startPosition}
        endHingePosition={endPosition}
        controlPoint={midPoint}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with scaledDown prop", () => {
    const startPosition = HingeGridPositionPoints.centerLeft;
    const endPosition = HingeGridPositionPoints.centerRight;
    const midPoint = getMidPointPosition(startPosition, endPosition);
    const { container } = render(
      <CurveLimbGroup
        startHingePosition={startPosition}
        endHingePosition={endPosition}
        controlPoint={midPoint}
        scaledDown
      />
    );
    expect(container).toMatchSnapshot();
  });
});
