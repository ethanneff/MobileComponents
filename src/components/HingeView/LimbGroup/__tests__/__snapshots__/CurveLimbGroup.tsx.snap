// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurveLimbGroup renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNSVGGroup
    fill={4278190080}
    fillOpacity={1}
    fillRule={1}
    font={Object {}}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={Array []}
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={null}
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth={1}
    vectorEffect={0}
    x={0}
    y={0}
  >
    <RNSVGPath
      d="M 550 880 Q 800 880 1050 880"
      fill={null}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={27}
      vectorEffect={0}
      x={0}
      y={0}
    />
    <RNSVGCircle
      cx={550}
      cy={880}
      fill={4287603630}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={22}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={12}
      vectorEffect={0}
      x={0}
      y={0}
    />
    <RNSVGCircle
      cx={1050}
      cy={880}
      fill={4287603630}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={22}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={12}
      vectorEffect={0}
      x={0}
      y={0}
    />
  </RNSVGGroup>
</View>
`;

exports[`CurveLimbGroup renders correctly with scaledDown prop 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNSVGGroup
    fill={4278190080}
    fillOpacity={1}
    fillRule={1}
    font={Object {}}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={Array []}
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={null}
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth={1}
    vectorEffect={0}
    x={0}
    y={0}
  >
    <RNSVGPath
      d="M 550 880 Q 800 880 1050 880"
      fill={null}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={27}
      vectorEffect={0}
      x={0}
      y={0}
    />
    <RNSVGCircle
      cx={550}
      cy={880}
      fill={4287603630}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={16}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={12}
      vectorEffect={0}
      x={0}
      y={0}
    />
    <RNSVGCircle
      cx={1050}
      cy={880}
      fill={4287603630}
      fillOpacity={1}
      fillRule={1}
      matrix={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
      opacity={1}
      originX={0}
      originY={0}
      propList={
        Array [
          "fill",
          "stroke",
          "strokeWidth",
        ]
      }
      r={16}
      rotation={0}
      scaleX={1}
      scaleY={1}
      skewX={0}
      skewY={0}
      stroke={4287603630}
      strokeDasharray={null}
      strokeDashoffset={null}
      strokeLinecap={0}
      strokeLinejoin={0}
      strokeMiterlimit={4}
      strokeOpacity={1}
      strokeWidth={12}
      vectorEffect={0}
      x={0}
      y={0}
    />
  </RNSVGGroup>
</View>
`;
